### DISCUSSION TITLE
How to use AND condition for roles and derived roles?

### DISCUSSION THREAD

**Discussion by afifurrohman-id** [cerbos/cerbos#2294] [Q&A] [answered]
Hello currently `roles` and `derived_roles` is **OR condition**, in resource policy either have roles OR dervied_roles is correct.
Even derived_roles and roles itself is also OR condition where either have role a / b OR derived_roles a / b is also correct.

But my usecase is i want 1 base derived roles: e.g for multi tenant system eval `attr` based on `tenant id` and each of this `tenant` have custom roles either static or derived_roles.

but still the context is based on this tenant.

but again is not possible currently, if i pass multiple role / derived roles that OR condition, so each tenant **not isolated** can access to other tenant, that not what we want.

current solution is using condition directly inside action itself. and just using 1 derived roles that eval the tenant id, but still not flexible enough.

Is there are better design or i'am fail on my implementation? 
any help will be appreciated, thanks...

**Comment by Sambigeara** [comment]
Hi,

To clarify: you want to be able to evaluate against rules defined on a per-tenant basis? If so, [scoped policies](https://docs.cerbos.dev/cerbos/latest/policies/scoped_policies) might be a good fit.

You define your resource policies with scopes mapped to your multi-tenant hierarchy (e.g. `acme.uk.hr` might represent the `HR` department in the `UK` arm of the `Acme` company 🤷‍♂️). You then pass the tenant specific scope in the request to the PDP.

The evaluation order is important here: the engine will look for matching rules at the outer-most scope (e.g. the tenant scope). If it doesn't find a matching rule, it will check the next scope up the chain, e.g. `acme.uk`, and continues until it either resolves all actions, or defaults to `DENY`. So ensure that any rules are explicitly defined in the tenant scopes for necessary `DENY`s etc.

Did I understand your case correctly, or have I missed something?

**Reply by afifurrohman-id** [reply]
Okay i may miss understand what scope do.

Here my case:
Resource policy: I have resource foo with root scope `` empty scope and action is empty 
Resource policy: i have resource foo with scope for tenant_a and tenant_b and **exactly** have same action (view) with required derived roles `customer` and `viewer` roles

Derived roles: customer derived roles just eval `R.attr.tenant in P.attr.tenants` and parent roles is user

Principal request: user1 have scope tenant_a with tenant_a in `tenants` attribute and user2 have scope tenant_b with tenant_b in `tenants` atrribute. both have user and viewer roles

Resource request: foo1 have scope tenant_a with tenant attr is tenant_a and foo2 have tenant_b  scope with tenant attr is tenant_b


However user1 can access resource foo2.

**Reply by afifurrohman-id** [reply]
I will give example if have enough time.

**Reply by Sambigeara** [reply]
Oh I think I might have misunderstood your original question.

In the simple case, a common pattern we see is determining tenant association through the use of runtime attributes (as you mentioned above).

If a resource carries an attribute `R.attr.tenant = tenant_1`, then it should only be available to principals which have `P.attr.tenants = ["tenant_1", ...]`. You've defined this within your derived role. Therefore, a resource rule that applies to that given derived role should prevent principals without `tenant_1` in `P.attr.tenants` from accessing the resource 🤔 

What am I missing? I might not understand your question -- apologies if not. A more concrete example would be great if you had time.

**Reply by afifurrohman-id** [reply]
Hi, sorry for late reply.

Here the playground: https://play.cerbos.dev/p/HjJE66c7d2c5hGuXYvFkgeSww9VGOGH4

Here the exported assets from playground (if you need it): https://europe-west1-cerbos-playground.cloudfunctions.net/download?instanceId=HjJE66c7d2c5hGuXYvFkgeSww9VGOGH4

**Reply by Sambigeara** [reply]
Ok so I think I confused matters by mentioning scopes in my original answer -- apologies for that.

The two (scoped) resource policies in your playground are functionally identical, so passing either `tenant_a` or `tenant_b` will result in the same evaluation and therefore the same output.

The reason for the following:

```yaml
  # TODO: why is allowed?
- principal: tenant_a_user1_customer
  resource: tenant_b_foo
  actions:
    view: EFFECT_DENY
```

Is due to the principals carrying the `viewer` role (as you've previously mentioned, `roles` and `derivedRoles` are OR'd when determining whether a rule applies to a principal or not). Removing the `roles: ["viewer"]` section from the resource policy rules results in your tests passing.

That said, scopes are unnecessary here. We can remove the scoped resource policies and consolidate into a single base level resource policy, and rely entirely on runtime attributes (via your derived role) to infer tenant association. I've demonstrated that in this playground example: https://play.cerbos.dev/p/Dk9xJs9vM1n66cede69oQYcOBhjP7uPb

**Reply by afifurrohman-id** [reply]
Thankyou for explanation.

I see, if i can simplify the concept, scope just like an **inheritance** for action right?

**Reply by Sambigeara** [reply, marked as answer]
> scope just like an inheritance for action right?

Actually, not quite! When you pass a `scope` in the request, the evaluation engine will start from the outer-most scope and work up the hierarchy until all actions are resolved, or you run out of policies (and end up with `DENY`s).

So, for example: for the resource scope `acme.hr.uk`, the engine will first retrieve the policy with scope `acme.hr.uk` and attempt to evaluate against the actions passed in the request. If not all actions are resolved in that first policy, it'll then move up one level and do the same against `acme.hr`, and then again (if necessary) against `acme` and then the default `""`. If all actions are resolved at a given scope level, the traversal finishes and the results are returned.

There are definitely similarities to inheritance, but also, unrelated (i.e. not overridden) rules can result in explicit evaluation outcomes, and so it muddies the water of how some people might expect "inheritance" to behave.

It's best to think of it as a chain of independently evaluated policies, rather than as inheritance.

**Reply by afifurrohman-id** [reply]
I see, really good explanation and make me more easy to understand.

Thankyou, really helpful
